### ISSUE TITLE
Unusually high sample KL

### ISSUE THREAD

**Issue by maitchison** [thinking-machines-lab/tinker-feedback#22]
I'm doing RL fine-tuning on OSS-120B, and finding that my training runs start with a sampling KL of 0.05, and within 20 steps, get to 0.1. This seems very high, and is certainly higher than the recommended limit of 0.01. 

I'm just using the provided `do_sync_training` from [train.py](https://github.com/thinking-machines-lab/tinker-cookbook/blob/0adfc25b7e1ea99d3ec4aab311516fab8c5e5120/tinker_cookbook/rl/train.py). 

Could this indicate an issue with the rollout or training servers (e.g. some kind of mismatch in the kernels)? Is there anything I can do on my end to reduce the sample KL?

**Comment by maitchison** [comment]
Thanks, Opher, here are the details.

* We're using cookbook, with a slightly modified copy of `do_sync_training` (just to add some logging)
* `ac_tokens` are around 2k, and `ob_tokens` around 6k
* `lr` varies, but `1e-4` in this case. We also see this with lower learning rates (e.g. `1e-5`). 
* `groups_per_batch` is 64
* `group_size` is 16
* renderer is the standard OSS one with medium effort
* It's a custom task, with policy entropy of around 0.5 (but rises to 0.6, and sometimes to 1.2)
* We see the same issue for `groups_per_batch=32` and `groups_per_batch=128`, although groups_per_batch=32 eventually degrades to a sample_kl of `0.14`!

**Comment by opherlieber** [comment]
Hey, not sure if related to your issue, but just an update that I've noticed that in general, gpt-oss has very bad KL initially on docs/formats its unfamiliar with. This stems from the fact it has very high NLL/logprobs for this data and we are calculating 
KL on the logprobs themselves (for example if you give it some wikipedia docs you'll see in the order of ~10x higher NLL then most other instruct models, e.g. around 3-4), and since there is always some noise/variance in logits/logprobs when calculating them concurrently, this can result in 100x higher 'KL' just from the precision noise since the KL formula is squared error on the logprobs.
So, I'd recommend checking what the logprobs/NLL are in your case, if they are high (e.g. 2-3-4 vs 10x less than that), it might make sense and not be an actual issue and eventually stabilize (Can also try some SFT on your data before RL in this case if that's an option).
The task I checked for example is DeepMath which is likely more in the initial output distribution of gpt-oss.

**Comment by maitchison** [comment]
> Thanks for the info! Is it 0.003 in Qwen3-30B-A3B vs <=0.02 in gpt-oss? How does NLL look on the prompt for the Qwen?

NLL For training on ac tokens is `0.3`, and for ob tokens is `1.67`.
